audioipc/src/shm.rs

```diff
@@ -117,39 +117,48 @@ impl SharedMemWriter {
 }
 
 pub struct SharedMemMutSlice {
-    view: MmapViewSync,
+    mmap: Arc<UnsafeCell<MmapMut>>,
 }
 
 impl SharedMemMutSlice {
     pub fn from(file: &File, size: usize) -> Result<SharedMemMutSlice> {
-        let mmap = Mmap::open(file, Protection::ReadWrite)?;
+        let mmap = unsafe { MmapOptions::new().map_mut(file)? };
         assert_eq!(mmap.len(), size);
-        let view = mmap.into_view_sync();
-        Ok(SharedMemMutSlice { view })
+        let mmap = Arc::new(UnsafeCell::new(mmap));
+        Ok(SharedMemMutSlice { mmap })
     }
 
     pub fn get_mut_slice(&mut self, size: usize) -> Result<&mut [u8]> {
         if size == 0 {
             return Ok(&mut []);
         }
         // TODO: Track how much is in the shm area.
-        if size <= self.view.len() {
-            let buf = unsafe { &mut self.view.as_mut_slice()[..size] };
+        if size <= self.inner().len() {
+            let buf = &mut self.inner_mut()[..size];
```

Comment:
If `UnsafeCell` is just used for breaking the mutability, we can replace `mmap: Arc<UnsafeCell<MmapMut>>` by `mmap: Arc<MmapMut>` and replace `let buf = &mut self.inner_mut()[..size]` by `let buf = &mut Arc::get_mut(&mut self.mmap).unwrap()[..size]`. Therefroe we can drop `UnsafeCell` and `inner*`.

Reply:
It's also used to signal that the underlying memory is potentially aliased (it is shared memory, after all), so I believe it's necessary to use UnsafeCell.

Reply:
I don't think I understand this correctly. What does the _aliased memory_ mean here? 

- Does the `Unsafe` need to be used in the function operation?
- Isn't `unsafe { MmapOptions::new().map_mut(file)? }` enough for signaling the unsafeness?
- Then we will have `SharedMemSlice { mmap: Arc<Mmap> }` and `SharedMemMutSlice { mmap: Arc<UnsafeCell<MmapMut>> }`.
  - Isn't`SharedMemSlice` is also shared memory or aliased memory? Why we only use `UnsafeCell` for `SharedMemMutSlice` if `UnsafeCell` is used to signal that the underlying memory is potentially aliased?
  - The difference here is _mut_.  Why the _mut_ is unsafe here so we need `UnsafeCell` for it? For those unsafe function operations, shouldn't they be marked `unsafe fn`?
- IIRC, `SharedMemSlice`(client side) is a pair of `SharedMemWriter`(server side) for the callback's input data. `SharedMemMutSlice`(client side) is a pair of `SharedMemReader`(server side) for the callback's output data. `SharedMemMutSlice` seems safe since the memory is only written by the client and only read on the server. Is there third side that can operate on the memory?

Reply:
Arc::get_mut doesn't work here because the reference count is > 1.  CallbackServer owns the SharedMemMutSlice and clones it temporarily into the CpuPool thread for processing.

UnsafeCell can't be replaced with RefCell because we need the mutable borrow to outlive get_mut_slice().

We can rework the shm.rs API, but that is outside the scope of this PR.  Also note that my original set of changes were directly based on the memmap code we were using before.

Reply:
oh I see. Thanks for the explanation!